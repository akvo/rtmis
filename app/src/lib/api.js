import axios from 'axios';

export const config = {
  baseURL: null,
  headers: {
    'Content-Type': 'application/json',
  },
};

const API = () => {
  const getConfig = () => {
    let current = config;
    if (api.baseURL) {
      current = {
        ...config,
        baseURL: api.baseURL,
      };
    }
    return api?.token
      ? {
          ...current,
          headers: { ...config.headers, Authorization: `Bearer ${api.token}` },
        }
      : current;
  };
  return {
    get: (url, config = {}) => axios({ url, ...getConfig(), ...config }),
    post: (url, data, config = {}) =>
      axios({
        url,
        method: 'POST',
        data,
        ...{
          ...getConfig(),
          headers: {
            ...getConfig().headers,
            ...config?.headers,
          },
        },
      }),
    put: (url, data, config) => axios({ url, method: 'PUT', data, ...getConfig(), ...config }),
    patch: (url, data, config) => axios({ url, method: 'PATCH', data, ...getConfig(), ...config }),
    delete: (url) => axios({ url, method: 'DELETE', ...getConfig() }),
    setToken: (token) => {
      api.token = token;
    },
    setServerURL: (serverURL) => {
      api.baseURL = serverURL;
    },
    getConfig,
  };
};

const api = API();

export default api;
