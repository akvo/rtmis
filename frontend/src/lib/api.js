import axios from "axios";

export const config = {
  baseURL: "/api/v1/",
  headers: {
    "Content-Type": "application/json",
  },
};

const API = () => {
  const getConfig = () => {
    return api?.token
      ? {
          ...config,
          headers: {
            ...config.headers,
            Authorization: `Bearer ${api.token}`,
          },
        }
      : config;
  };
  return {
    get: (url, config = {}) => axios({ url, ...getConfig(), ...config }),
    post: (url, data, config = {}) =>
      axios({ url, method: "POST", data, ...getConfig(), ...config }),
    put: (url, data, config) =>
      axios({ url, method: "PUT", data, ...getConfig(), ...config }),
    patch: (url, data, config) =>
      axios({ url, method: "PATCH", data, ...getConfig(), ...config }),
    delete: (url) => axios({ url, method: "DELETE", ...getConfig() }),
    setToken: (token) => {
      api.token = token;
    },
  };
};

const api = API();

export default api;
