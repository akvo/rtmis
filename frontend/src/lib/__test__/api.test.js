import MockAdapter from "axios-mock-adapter";
import "@testing-library/jest-dom";
import api from "../api";
import axios from "axios";

const fakeToken = "eyJhbGciOiJIUzI1NiIsInR56IkpXVCxxxxxxxxxxx";

describe("lib/api", () => {
  test("test if token is being stored in the api and sending correct headers", async () => {
    const mock = new MockAdapter(axios);
    api.setToken(fakeToken);
    expect(api.token).toStrictEqual(fakeToken);

    mock.onGet("/something").reply((config) => {
      expect(config.baseURL).toEqual("/api/v1/");
      expect(config.headers).toEqual({
        Accept: "application/json, text/plain, */*",
        Authorization: `Bearer ${fakeToken}`,
        "Content-Type": "application/json",
      });
      return [200, {}];
    });

    await api.get("/something");
  });
});
