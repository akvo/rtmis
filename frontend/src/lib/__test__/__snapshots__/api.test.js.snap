// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/api snapshot api calls 1`] = `
Object {
  "delete": [Function],
  "get": [MockFunction],
  "patch": [Function],
  "post": [Function],
  "put": [Function],
  "setToken": [Function],
  "token": "eyJhbGciOiJIUzI1NiIsInR56IkpXVCxxxxxxxxxxx",
}
`;
